pkg/skaffold/deploy/kpt.go

```diff
@@ -80,12 +96,79 @@ func (k *KptDeployer) Cleanup(ctx context.Context, _ io.Writer) error {
 	return nil
 }
 
-func (k *KptDeployer) Render(ctx context.Context, out io.Writer, builds []build.Artifact, offline bool, filepath string) error {
-	return nil
+func (k *KptDeployer) Render(ctx context.Context, out io.Writer, builds []build.Artifact, _ bool, filepath string) error {
+	manifests, err := k.renderManifests(ctx, out, builds)
+	if err != nil {
+		return err
+	}
+
+	return outputRenderedManifests(manifests.String(), filepath, out)
+}
+
+func (k *KptDeployer) renderManifests(ctx context.Context, out io.Writer, builds []build.Artifact) (deploy.ManifestList, error) {
+	debugHelpersRegistry, err := config.GetDebugHelpersRegistry(k.globalConfig)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving debug helpers registry: %w", err)
+	}
+
+	manifests, err := k.runKpt(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("running kpt functions: %w", err)
+	}
+
+	if len(manifests) == 0 {
+		return nil, nil
+	}
+
+	manifests, err = manifests.ReplaceImages(builds)
+	if err != nil {
+		return nil, fmt.Errorf("replacing images in manifests: %w", err)
+	}
+
+	for _, transform := range manifestTransforms {
+		manifests, err = transform(manifests, builds, Registries{k.insecureRegistries, debugHelpersRegistry})
+		if err != nil {
+			return nil, fmt.Errorf("unable to transform manifests: %w", err)
+		}
+	}
+
+	return manifests.SetLabels(k.labels)
+}
+
+// runKpt does a dry run with the specified kpt functions (fn-path XOR image) against dir.
+// If neither fn-path or image are specified, functions will attempt to be discovered in dir.
+// An error is returned when both fn-path AND image are specified.
+func (k *KptDeployer) runKpt(ctx context.Context) (deploy.ManifestList, error) {
+	var manifests deploy.ManifestList
+
+	flags := []string{"--dry-run"}
```

Comment:
nit: can we add a comment saying `--dry-run` will put the `kpt fn run` output resource in stdout, otherwise ...